src/backends/r1cs/mod.rs

```diff
@@ -121,17 +142,14 @@ impl R1csBls12_381 {
         }
     }
 
-    // todo: how can we create this r1cs backend with different associated field types, but still using the same backend implementation? using macro?
-    /// So that we can get the associated field type, instead of passing it as a parameter here.
-    /// There are two fields supported by the snarkjs for r1cs: bn128 and ark_bls12_381.
-    /// Currently we are using ark_bls12_381
+    /// Returns the prime for snarkjs based on the curve field.
     fn prime(&self) -> BigInt {
-        BigInt::from_slice_native(Sign::Plus, Fr::characteristic())
+        BigInt::from_slice_native(Sign::Plus, F::characteristic())
```

Comment:
I don't think [characteristic](https://en.wikipedia.org/wiki/Characteristic_(algebra)) is the right function here btw. 

I think the problem of this function is that it assumes a `PrimeField` (which has a method to give you the prime) and not a `Field`

so my recommendation would be to expect a `F: BackendField + PrimeField` here and do `F::Params::MODULUS`